File: src/zros_cpp/service_client_manager.cc
//
// Created by pairs on 7/17/18.
//

#include <zros/service_client_manager.h>
#include <zros/singleton.h>
#include <sspdlog/sspdlog.h>
#include "impl/service_clients_impl.h"

namespace zros {

    ServiceClientManager::ServiceClientManager() {
        impl_ = std::make_shared<ServiceClientsImpl>();
        service_discovery_ = SingleTon<ServiceDiscovery>::Instance();
        service_discovery_->set_register_service_server_cb([this](const zros_rpc::ServiceServerInfo *serverInfo, zros_rpc::Status *status){
            impl_->registerServiceClient(serverInfo, status);
        });
        service_discovery_->set_unregister_service_server_cb([this](const zros_rpc::ServiceServerInfo *serverInfo, zros_rpc::Status *status) {
            impl_->unregisterServiceClient(serverInfo, status);
        });


    }

    bool ServiceClientManager::registerClient(const std::shared_ptr<zros::IServiceClient> client) {
        // 1. register to master first
        bool ok = service_discovery_->addServiceClient(client);
        if (ok) {
            SSPD_LOG_INFO << "register client " << client->get_service_name() << " to master success";
            // 2. register to memory
            impl_->registerClient(client);
            return true;
         }
        SSPD_LOG_WARNING << "register client " << client->get_service_name() << " to master failed";
        return false;
    }

    bool ServiceClientManager::unregisterClient(const std::string &service_name) {
        // todo
        return false;
    }

    std::shared_ptr<zros_rpc::ServiceResponse>
    ServiceClientManager::call(const std::string &service_name, const std::string &content, const std::string &cli_info,
                               int timeout_mseconds) {
        return impl_->call(service_name, content, cli_info, timeout_mseconds);
    }
}

File: src/zros_python/service_client_manager.py
# -*- coding: utf-8 -*-
""" The service client management

data: 2018-08-01
author: pairs
"""

import zros_python.impl.service_clients_impl as impl
import zros_python.zros as zros
import logging

logger = logging.getLogger(__name__)


class ServiceClientManager(object):
    """

    """
    def __init__(self):
        self._impl = impl.ServiceClientsImpl()
        self._gsd = zros.get_service_discovery()
        self._gsd.set_register_service_server_cb(self._impl.register_service_client)
        self._gsd.set_unregister_service_server_cb(self._impl.unregister_service_client)

    def register_client(self, client):
        self._impl.register_client(client)
        ok = self._gsd.add_service_client(client)
        if ok:
            logger.info(u'register client %s to master success', client.get_service_name())
            return True
        else:
            self._impl.unregister_service_client(client.get_service_name())
            logger.warning(u'register client %s to master failed', client.get_service_name())
            return False

    def unregister_client(self, service_name):
        pass

    def call(self, service_name, content, cli_info, timeout):
        return self._impl.call(service_name, content, cli_info, timeout)
